source/audio/appsVolume.py

```diff
@@ -0,0 +1,159 @@
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2024 NV Access Limited
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+
+import config
+import globalVars
+from logHandler import log
+import nvwave
+from pycaw.utils import AudioSession
+import ui
+from dataclasses import dataclass
+from threading import Lock
+from config.featureFlagEnums import AppsVolumeAdjusterFlag
+from typing import NamedTuple
+from .utils import AudioSessionCallback, DummyAudioSessionCallback
+
+
+class VolumeAndMute(NamedTuple):
+	volume: float
+	mute: bool
+
+
+_appVolumesCache: dict[int, VolumeAndMute] = {}
+_appVolumesCacheLock = Lock()
+_activeCallback: DummyAudioSessionCallback | None = None
+
+
+def initialize() -> None:
+	state = config.conf["audio"]["applicationsVolumeMode"]
+	if state == AppsVolumeAdjusterFlag.MUTED:
+		state = AppsVolumeAdjusterFlag.ENABLED
+		config.conf["audio"]["applicationsVolumeMode"] = state
+	volume = config.conf["audio"]["applicationsSoundVolume"]
+	_updateAppsVolumeImpl(volume / 100.0, state)
+
+
+def terminate():
+	global _activeCallback
+	if _activeCallback is not None:
+		_activeCallback.unregister()
+		_activeCallback = None
+
+
+@dataclass(unsafe_hash=True)
+class VolumeSetter(AudioSessionCallback):
+	volumeAndMute: VolumeAndMute | None = None
+
+	def getOriginalVolumeAndMute(self, pid: int) -> VolumeAndMute:
+		try:
+			with _appVolumesCacheLock:
+				originalVolumeAndMute = _appVolumesCache[pid]
+				del _appVolumesCache[pid]
+		except KeyError:
+			originalVolumeAndMute = VolumeAndMute(volume=1.0, mute=False)
+		return originalVolumeAndMute
+
+	def onSessionUpdate(self, session: AudioSession) -> None:
+		pid = session.ProcessId
+		simpleVolume = session.SimpleAudioVolume
+		with _appVolumesCacheLock:
+			if pid not in _appVolumesCache:
+				_appVolumesCache[pid] = VolumeAndMute(
+					volume=simpleVolume.GetMasterVolume(),
+					mute=simpleVolume.GetMute(),
+				)
+		if pid != globalVars.appPid:
+			simpleVolume.SetMasterVolume(self.volumeAndMute.volume, None)
+			simpleVolume.SetMute(self.volumeAndMute.mute, None)
+
+	def onSessionTerminated(self, session: AudioSession) -> None:
+		pid = session.ProcessId
+		simpleVolume = session.SimpleAudioVolume
+		originalVolumeAndMute = self.getOriginalVolumeAndMute(pid)
+		try:
+			simpleVolume.SetMasterVolume(originalVolumeAndMute.volume, None)
+			simpleVolume.SetMute(originalVolumeAndMute.mute, None)
+		except Exception:
+			log.exception(f"Could not restore master volume of process {pid} upon exit.")
+

```

Comment:
The rabbit proposes to do nothing if previous volumeAndMute settings cannot be found. Not sure if this is a better solution. Currently we restore volume to 100% if previous volume not found. I also must say that this is just a safety thing - I didn't encounter this situation in my own testing and don't expect this will happen during normal operation. So I believe current logic is good enough and not worth changing.